### PR TITLE
Implement an API client for pecl.php.net

### PR DESCRIPTION
**Pull request by NiR-**
The API exposed at pecl.php.net is not documented (or the documentation
is well hidden). As such, endpoints used in this commit come from PEAR
codebase. The client has 6 methods:

* ListPackages(), based on `/p/packages.xml` ;
* ListPackagesInCategory(), based on `/c/{category}/packages.xml` ;
* DescribePackage(), based on `/p/{packageName}/info.xml` ;
* ListReleases(), based on `/r/{packageName}/allreleases.xml` ;
* DescribeRelease(), based on `/r/{packageName}/{version}.xml` ;
* DownloadRelease(), uses the PartialURI returned by DescribeRelease()
  and appends `.tgz` extension to it ;

Both Describe methods use custom types to represent their respective
data. Moreover, ListReleases uses a custom type named PackageReleases,
which supports sorting.

### CODE REVIEW

File: peclapi/client.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+type releaseList struct {
+	Releases []release `xml:"r"`
+}
+
+type release struct {
+	Version   string `xml:"v"`
+	Stability string `xml:"s"`
+}
+
+// PackageReleases is the list of releases associated to their stabilit for a
```
Comment (NiR-) [inline peclapi/client.go]:
```suggestion
// PackageReleases is the list of releases associated to their stability for a
```

File: peclapi/client.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+	Summary      string `xml:"s"`
+	Description  string `xml:"d"`
+	ReleaseDate  string `xml:"da"`
+	ReleaseNotes string `xml:"n"`
+	PartialURI   string `xml:"g"`
+	PackageXML   string `xml:"x"`
+}
+
+// DownloadRelease downloads a given package release and returns an
+// io.ReadCloser from which a tar can be read. An error is returned if the HTTP
```
Comment (NiR-) [inline peclapi/client.go]:
```suggestion
// io.ReadCloser from which a gzip can be read. An error is returned if the HTTP
```

File: peclapi/client.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+// request fails, if a bad status code is returned or if the downloaded file is
+// not an application/x-gzip.
+func (c Client) DownloadRelease(release Release) (io.Reader, error) {
+	if release.PartialURI == "" {
+		return nil, xerrors.Errorf("empty PartialURI")
+	}
+
+	url := fmt.Sprintf("%s.tgz", release.PartialURI)
```
Comment (NiR-) [inline peclapi/client.go]: Add a comment about this sprintf

File: peclapi/client.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+	"sort"
+
+	"github.com/mcuadros/go-version"
+	"golang.org/x/text/encoding/ianaindex"
+	"golang.org/x/text/transform"
+	"golang.org/x/xerrors"
+)
+
+// NewClient creates a new HTTP-based client for the API hosted at
+// pecl.php.net/redis/. It takes ClientOpt as arguments. These could be used to
```
Comment (nm2107) [inline peclapi/client.go]: `redis` => `rest` ?

File: peclapi/client.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+func (s versionSlice) Len() int {
+	return len(s)
+}
+
+func (s versionSlice) Less(i, j int) bool {
+	return version.Compare(s[i], s[j], "<")
+}
+
+func (s versionSlice) Swap(i, j int) {
```
Comment (nm2107) [inline peclapi/client.go]: what's the need to swap versions ?

Comment (NiR-) [reply peclapi/client.go]: `Len()`, `Less()` and `Swap()` methods actually implements [`Sort.Interface`](https://pkg.go.dev/sort?tab=doc#Interface) to make `versionSlice` compatible with `sort.Sort()` function :winl:

**Comment by NiR-** [comment]
> lets hope they won't change their definitions grin

If the `last updated at` comments in definitions files are correct, those file haven't been touched since 2005 :smile:

**Pull request merged**